### ISSUE TITLE
The kubelet /stats/summary is too expensive on windows

### ISSUE THREAD

**Issue by wawa0210** [kubernetes/kubernetes#82522]
**What happened**:

Collect host and container information via kubelet stats/summay. By looking the kubelet code ,it's collected by cadvisor and the stats/summary request is kept within 50 ms. 

But the implementation in the Windows environment is very cumbersome。First it will get all the pods managed by kubelet ，and call docker stats request one by one (very very expensive)

This is my set of test data

The pod is very sample just the `mcr.microsoft.com/windows/nanoserver:1903` and loop ping

|  pod nums   | stats/summay times  |
|  ----  | ----  |
| 5  | 10s |
| 10  | 20s |
| 15  | 30.50s |
| 20  | 41.38s |

On the Linux platform, the time cost will not change too much and remain within 100ms.
This is unacceptable on windows .

By studying the kubelet code, the main time consuming is to request the stats of each container through docker, which takes up almost 95% of the entire request time.

https://github.com/kubernetes/kubernetes/blob/1888745c3db2432a613b2649686ee6bf5c9d2fa1/pkg/kubelet/dockershim/docker_stats_windows.go#L34

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:

Deploy 5 to 10 pods to the windows host then request kubelet 10255 stats/summay 

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use kubectl version): 1.15.3
- Windows server version : windows server 2019 17763.678
- Docker version : docker ee 18.09.6

/sig node

**Comment by rhockenbury** [comment]
This has been reported previously in https://github.com/kubernetes/kubernetes/issues/75752 and https://github.com/kubernetes/kubernetes/issues/74991

**Comment by Random-Liu** [comment]
We are working on adding the support in containerd, and we'll get stats from the hcsshim. https://github.com/containerd/cri/issues/1275

If `docker stats` is where the overhead comes from, we can optimize it in the following ways:
1) Actually we only need very minimal stats about a container in CRI https://github.com/kubernetes/cri-api/blob/master/pkg/apis/runtime/v1alpha2/api.proto#L1248, a lot of stats are not needed. If it turns out that avoid collecting unnecessary stats can significantly reduce the request time, we can probably find a way to support it.
2) If it turns out that stats collection is just slow, we can consider:
* Option 1: Collecting stats in parallel, if stats collection doesn't take much resource, and most times it is just block and wait;
* Option 2: Caching stats and update in background. This is what we do for snapshot stats for linux, because snapshot stats collection is slow and resource consuming. https://github.com/containerd/cri/blob/master/pkg/server/snapshots.go#L71

**Comment by Random-Liu** [comment]
I tested the stats on containerd on windows, and it is fairly cheap.
```
$ crictl ps -q
5e7d99e0147e76d3c9c9f75206c06f6e59f9924c2a271158a6ff9197010f56bb
dfe41d01b7728fcfd25302725abd550960277b8875333fb86eb0b6b5d318a4ec
4735691009c5f9bdacf409d3e7b6b81af24dc45d270cb8c781c9ed946f4cfe74
0e70a82a114f3b38610da0a302d09a8b6aefac9f58fe7aad7eba856b5003841a
58073101c471cbc35c926fd0a3ef5a0caddd94646de7fca87d56d2572d591aac
af8a70bd05b8b37161c4f8bc523b2999e756993d0ea642c23bb94ed364f4a139
10a88cd65c75d386e579220f70baa5b0075f515d35e065a3aa86eb9055ea0a16
f49f5fe91d7b11e856d90c21d6e28aec21add8d64bca247b2ae5d2c1a684ae1f
85d5e2d90194be703a91c9c1609eca814a64556aa7eef0c5a01559912661171f
fca0e3d79ea54916de7fc0e44b2f4e8c3d7b121ed3c4f3ecf5cebd848d69e633

$ time crictl stats -o json
....
real    0m0.094s
user    0m0.000s
sys     0m0.015s
```

Please note that the writable layer size is not supported yet in containerd.

@wawa0210 Are you sure `GetContainerStats` takes most of the time?

I saw that in Docker we use `InspectContainerWithSize` to get the writable layer size. That operation can potentially take a long time based on the underlying implementation. And for linux, we cache that size instead of collecting on demand.

**Comment by wawa0210** [comment]
> I tested the stats on containerd on windows, and it is fairly cheap.
> 
> ```
> $ crictl ps -q
> 5e7d99e0147e76d3c9c9f75206c06f6e59f9924c2a271158a6ff9197010f56bb
> dfe41d01b7728fcfd25302725abd550960277b8875333fb86eb0b6b5d318a4ec
> 4735691009c5f9bdacf409d3e7b6b81af24dc45d270cb8c781c9ed946f4cfe74
> 0e70a82a114f3b38610da0a302d09a8b6aefac9f58fe7aad7eba856b5003841a
> 58073101c471cbc35c926fd0a3ef5a0caddd94646de7fca87d56d2572d591aac
> af8a70bd05b8b37161c4f8bc523b2999e756993d0ea642c23bb94ed364f4a139
> 10a88cd65c75d386e579220f70baa5b0075f515d35e065a3aa86eb9055ea0a16
> f49f5fe91d7b11e856d90c21d6e28aec21add8d64bca247b2ae5d2c1a684ae1f
> 85d5e2d90194be703a91c9c1609eca814a64556aa7eef0c5a01559912661171f
> fca0e3d79ea54916de7fc0e44b2f4e8c3d7b121ed3c4f3ecf5cebd848d69e633
> 
> $ time crictl stats -o json
> ....
> real    0m0.094s
> user    0m0.000s
> sys     0m0.015s
> ```
> 
> Please note that the writable layer size is not supported yet in containerd.
> 
> @wawa0210 Are you sure `GetContainerStats` takes most of the time?
> 
> I saw that in Docker we use `InspectContainerWithSize` to get the writable layer size. That operation can potentially take a long time based on the underlying implementation. And for linux, we cache that size instead of collecting on demand.

Thanks for the reply, yes, I am sure `GetContainerStats` is the most time consuming.
Because in our own environment, these stats information is not required, I commented this code, and the interface response time is almost the same as Linux. 
I later tested the return time of the docker stats interface separately. It is very slow in linux and windows. There is no difference. The response time of a single request is close to 1s.
As you said, kubernetes caches the Linux response environment.

**Comment by yliaog** [comment]
I timed various calls in:
https://github.com/kubernetes/kubernetes/blob/2bb35caa18d0c2c3ade2d89c41158dfbfdbc54b2/pkg/kubelet/dockershim/docker_stats_windows.go#L28

W1026 05:01:07.787724    2940 docker_stats_windows.go:82] Info: 9.7872ms
W1026 05:01:07.787724    2940 docker_stats_windows.go:83] GetContainerStats: 2.2006699s
W1026 05:01:07.787724    2940 docker_stats_windows.go:84] InspectContainerWithSize: 34.1736ms
W1026 05:01:07.787724    2940 docker_stats_windows.go:85] ContainerStatus: 9.7692ms

Indeed, GetContainerStats is very slow ...

**Comment by yliaog** [comment]
run 'docker stats' directly show about the same amount time as that from kubelet docker_stats_windows.go
PS > Measure-Command {docker stats 9a476754d43d --no-stream}
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 2
Milliseconds      : 427
Ticks             : 24273497
TotalDays         : 2.80943252314815E-05
TotalHours        : 0.000674263805555556
TotalMinutes      : 0.0404558283333333
TotalSeconds      : 2.4273497
TotalMilliseconds : 2427.3497

'docker stats' output total 686 bytes:
PS > docker stats 9a476754d43d --no-stream
CONTAINER ID        NAME                                                                         CPU %               PRIV WORKING SET    NET I/O             BLOCK I/O
9a476754d43d        k8s_POD_iis-744bbf99d-4rdgr_default_9f464653-6da4-4efb-b483-50974f74a414_0   0.70%               23.33MiB            587kB / 109kB       41.5MB / 7.64MB

**Comment by Random-Liu** [comment]
OK. So it is something in `docker stats` really slow.

And we don't have that issue in containerd.

**Comment by yliaog** [comment]
'docker stats' on all containers takes about 1s longer than 'docker stats' on *one* container. so instead of polling for 'docker stats' for each container one by one, it's a possible optimization to get all container stats in one call.

PS > Measure-Command {docker stats --no-stream}
Days              : 0
Hours             : 0
Minutes           : 0
Seconds           : 3
Milliseconds      : 749
Ticks             : 37494197
TotalDays         : 4.33960613425926E-05
TotalHours        : 0.00104150547222222
TotalMinutes      : 0.0624903283333333
TotalSeconds      : 3.7494197
TotalMilliseconds : 3749.4197

**Comment by cpuguy83** [comment]
The reason for the slowness is the stats endpoint in Docker does 2 collections with a 1s sleep in between... and the sleep happens at the beginning of the loop.

Docker 19.03.7 has been changed such that the sleep now occurs at the end of the loop, which gives a pretty significant speed up.

In Docker 20.03 (version not finalized), we should have a new query param, "one-shot=true", which will do a single collection.

**Comment by cpuguy83** [comment]
Fix sleep:
https://github.com/moby/moby/pull/40549

New param:
https://github.com/moby/moby/pull/40478

**Comment by PatrickLang** [comment]
Fixed in https://github.com/kubernetes/kubernetes/pull/87730

If you have another case where this is slow on 1.18, please open up a new issue and let us know which api calls are slowing down.

/close
